src/dataflow-types/src/client/controller/compute.rs

```diff
@@ -587,13 +590,15 @@ where
                 .collection_mut(*id)
                 .expect("Reference to absent collection");
 
+            let old_write_frontier = collection.write_frontier.frontier().to_owned();
             collection
                 .write_frontier
                 .update_iter(changes.clone().drain());
 
-            let mut new_read_capability = collection
-                .read_policy
-                .frontier(collection.write_frontier.frontier());
+            let mut new_read_capability = collection.read_policy.frontier(
+                collection.write_frontier.frontier(),
+                old_write_frontier.borrow(),
+            );
```

Comment:
It seems possible that this could fire for an empty `changes`, for which the old and new frontiers are the same. It seems like we might want to avoid that, in the spirit of keeping the "semantically" old frontier, to ensure that we have at least one frontier available for immediate querying.

Reply:
Good point. I added a `changes.is_empty() -> continue` check here but I'm not sure of its correctness. Is that the best way to do this?